src/linear_algebra/contraction.lean

```diff
@@ -51,4 +71,85 @@ variables {R M N}
+  (bM : basis m R M) (bN : basis n R N) (j : m) (i : n) :
+    to_matrix bM bN (dual_tensor_hom R M N (bM.coord j ⊗ₜ bN i)) = std_basis_matrix i j 1 :=
+begin
+  ext i' j',
+  by_cases hij : (i = i' ∧ j = j');
+  simp [linear_map.to_matrix_apply, finsupp.single_eq_pi_single, hij],
+  rw [and_iff_not_or_not, not_not] at hij, cases hij; simp [hij],
+end
+
+/-- `hom_dual_tensor` is right inverse to `dual_tensor_hom` -/
+@[simp]
+lemma dual_tensor_hom_hom_dual_tensor_apply (f : M →ₗ[R] N):
+  (dual_tensor_hom R M N) ((hom_dual_tensor R M N b) f) = f :=
+begin
+  ext m, nth_rewrite_rhs 0 ←basis.sum_repr b m,
+  simp [hom_dual_tensor],
+end
+
+/-- `hom_dual_tensor` is right inverse to `dual_tensor_hom` -/
+lemma dual_tensor_hom_hom_dual_tensor :
+  (dual_tensor_hom R M N) ∘ₗ (hom_dual_tensor R M N b) = id :=
+by { ext f, simp }
+
+/-- `hom_dual_tensor` is left inverse to `dual_tensor_hom` -/
+lemma hom_dual_tensor_dual_tensor_hom :
+  (hom_dual_tensor R M N b) ∘ₗ (dual_tensor_hom R M N) = id :=
+begin
+  apply curry_injective,
+  apply basis.ext b.dual_basis,
+  intro i, ext n,
+  suffices H : ∑ (x : ι), (finsupp.single x (1:R)) i • b.coord x ⊗ₜ[R] n = b.coord i ⊗ₜ[R] n,
+  { simp [hom_dual_tensor_apply, -finsupp.single_one_smul, H] },
+  simp,
+end
+
+noncomputable
+def dual_tensor_hom_equiv_of_basis :
+  (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N :=
+linear_equiv.of_linear (dual_tensor_hom R M N) (hom_dual_tensor R M N b)
+  (dual_tensor_hom_hom_dual_tensor b) (hom_dual_tensor_dual_tensor_hom b)
+
+lemma dual_tensor_hom_equiv_of_basis_eq : dual_tensor_hom_equiv_of_basis b =
+  (dual_tensor_hom_equiv_of_basis c : (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N) :=
+by {ext x m, simp [dual_tensor_hom_equiv_of_basis] }
+
+end
+
+open finite_dimensional
+variables (R : Type u) (M N : Type v)
+variables [field R] [add_comm_group M] [add_comm_group N] [module R M] [module R N]
+variables [finite_dimensional R M]
+
+/-- `dual_tensor_hom` is an equivalence -/
+noncomputable def dual_tensor_hom_equiv : (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N :=
+dual_tensor_hom_equiv_of_basis (fin_basis R M)
+
+lemma coe_dual_tensor_hom_equiv :
+  (dual_tensor_hom_equiv R M N : (module.dual R M) ⊗[R] N → M →ₗ[R] N) = dual_tensor_hom R M N :=
+rfl
+
+@[simp] lemma coe_dual_tensor_hom_equiv_to_linear :
+  (dual_tensor_hom_equiv R M N : (module.dual R M) ⊗[R] N →ₗ[R] M →ₗ[R] N)
+  = dual_tensor_hom R M N := rfl
+
+@[simp] lemma dual_tensor_hom_equiv_apply (f : (module.dual R M) ⊗[R] N):
+  (dual_tensor_hom_equiv R M N) f = (dual_tensor_hom R M N) f := rfl
+
+/-- `hom_dual_tensor` is equal to the inverse of `dual_tensor_hom`, no matter the basis used-/
+lemma dual_tensor_hom_equiv_symm
+  {ι : Type} [fintype ι] [decidable_eq ι] (b : basis ι R M) :
+  ⇑(dual_tensor_hom_equiv R M N).symm = hom_dual_tensor R M N b :=
+begin
+  ext x,
+  simp [dual_tensor_hom_equiv, dual_tensor_hom_equiv_of_basis,
+       dual_tensor_hom_equiv_of_basis_eq _ b],
+end
```

Comment:
It's tempting to just do
```suggestion
/-- `dual_tensor_hom` is an equivalence -/
@[simp] noncomputable def dual_tensor_hom_equiv : (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N :=
dual_tensor_hom_equiv_of_basis (fin_basis R M)
```
rather than repeating lemmas.